File: data/src/main/java/pl/mobly/data/SystemApiImpl.java
package pl.mobly.data;


import android.content.pm.ApplicationInfo;
import android.util.Log;

import java.util.List;

import io.reactivex.Observable;
import pl.mobly.domain.DomainApplicationInfo;
import pl.mobly.domain.SystemApi;

import static android.content.ContentValues.TAG;

public class SystemApiImpl implements SystemApi {
    private PackageManagerResource packageManagerResource;

    public SystemApiImpl(PackageManagerResource packageManagerResource) {

        this.packageManagerResource = packageManagerResource;
    }

    @Override
    public Observable<List<DomainApplicationInfo>> loadAppsList() {
        Observable.fromArray(
                AppInfoMapper
                        .mapToDomain(packageManagerResource.loadInstalledApps()))
                .doOnComplete(() -> {})
                .doOnNext(domainApplicationInfos -> {
                    Log.i(TAG, "loadAppsList: " + domainApplicationInfos.toString());
                }).subscribe();
        return null;
    }
}

File: data/src/main/java/pl/mobly/data/SystemApiImpl.kt
package pl.mobly.data

import android.content.pm.PackageManager
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import pl.mobly.domain.DomainApplicationInfo
import pl.mobly.domain.SystemApi


class SystemApiImpl(private val packageManagerResource: PackageManagerResource, private val packageManager: PackageManager) : SystemApi {

    override fun loadAppsList(): Observable<List<DomainApplicationInfo>> {
        return Observable.fromArray(
                AppInfoMapper
                        .mapToDomain(packageManagerResource.loadInstalledApps(), packageManager))
                .observeOn(Schedulers.io())
                .subscribeOn(AndroidSchedulers.mainThread())

    }
}
